### PR TITLE
Broadcast transactions only after their timelock is up

### PR DESCRIPTION
**Pull request by TheBlueMatt**
Built on #928, this avoids broadcasting transactions too early resulting in us leaning on users to do rebroadcasting. Instead, we wait until timelocks are up to broadcast.

### CODE REVIEW

File: lightning/src/chain/channelmonitor.rs
Comment (ariard) [inline lightning/src/chain/channelmonitor.rs:1915]: nit: you can rename `res` to `htlc_success_txn`

File: lightning/src/chain/onchaintx.rs
Comment (ariard) [inline lightning/src/chain/onchaintx.rs:410]: Duplicate the above logger, as it's already registered, maybe mentioning they're registered in the special locktime buffer.

File: lightning/src/chain/onchaintx.rs
Comment (ariard) [inline lightning/src/chain/onchaintx.rs:447]: I think I would gather all preprocessing checks (aggregation, delayed-claiming) in the same for loop above and the restoring loop as the first thing in `update_claims_view`. Post-anchor, we might be interested to aggregate our HTLC-timeout with an equal cltv_expiry and it won't work with this code organization.

File: lightning/src/chain/onchaintx.rs
Comment (ariard) [inline lightning/src/chain/onchaintx.rs:448]: Can we broadcast delayed-transactions scenario with the following reorg-scenario ?
* delayed-transactions are cached in `locktimed_packages`
* height expiration is reached, we generate and broadcast delayed transactions
* chain is reorg-out, delayed transactions are evicted from mempools
* before syncing to new tip, we receive a preimage offchain and enter into `update_claims_view`

I don't think we can regenerate and broadcast a non-propagating delayed transaction as regeneration timer are height-based. In the future, if we switch regeneration timer to mempool feeration variations, we might broadcast such delayed transactions.

Otherwise, I can't think of other edgy reorg cases.

File: lightning/src/chain/package.rs
Comment (ariard) [inline lightning/src/chain/package.rs:603]: nit: *outputs an "only valid for now (legacy channel types)"

File: lightning/src/chain/package.rs
Comment (ariard) [inline lightning/src/chain/package.rs:391]: I think input_height should be called "output_conf_height". I concede that height original isn't really clear, but it avoids at least letting think that this timelock are only CSV ones.

Maybe rename `amt` to `absolute_height`?

And precise the +1 is to follow mempool rules of evaluating timelocks, otherwise a newcomer can be easily confused on this.

File: lightning/src/chain/package.rs
Comment (ariard) [inline lightning/src/chain/package.rs:613]: nit: min_locktime? I would rather call the maximum locktime `std::u32::MAX`

File: lightning/src/chain/onchaintx.rs
Comment (TheBlueMatt) [reply lightning/src/chain/onchaintx.rs:448]: In that case don't we still generate the relevant commitment + timeout/claim transactions when the block comes back in when we sync to the new tip? Either way we should bump after a few blocks if we haven't confirmed by then.

File: lightning/src/chain/package.rs
Comment (TheBlueMatt) [reply lightning/src/chain/package.rs:603]: Hmm, not sure what you're suggesting? I'll let you update the docs for anchor in the anchor PRs.

File: lightning/src/chain/package.rs
Comment (TheBlueMatt) [reply lightning/src/chain/package.rs:613]: Hmm? This is calculating the maximum locktime of all included inputs, so its still a maximum?

**Comment by TheBlueMatt** [comment]
Rebased after merge of #928 and addressed comments.

File: lightning/src/chain/onchaintx.rs
Comment (ariard) [reply lightning/src/chain/onchaintx.rs]: Clearly, not a big deal. It's a just a edge-case about new feature added by this PR in case of a hypothetical future change. If we switch or add the option of mempool-based timing we'll have to split our hairs on few others things around OnchainTxHandler.

(Thought I still think mempool fluctuations-based timing might be more more interesting for fee savings than dumb rebroadcast after blocks X, if you're not near timelocks expiration)

File: lightning/src/chain/package.rs
Comment (ariard) [reply lightning/src/chain/package.rs:617]: Ah gotcha, I was thinking about the max absolute value that `nLocktime` might be. `max_default_locktime` would be a bit clearer.

**Review by ariard — COMMENTED** [review]
Code Review ACK 2f755b1

I went through all the fixes commit, globally satisfied with new comments. I had a look on test changes they look good overall.

File: lightning/src/ln/functional_test_utils.rs
Comment (jkczyz) [inline lightning/src/ln/functional_test_utils.rs:215]: Why does this need to be an `Arc`?

File: lightning/src/util/test_utils.rs
Comment (jkczyz) [inline lightning/src/util/test_utils.rs:205]: Likewise, why is an `Arc` needed?

File: lightning/src/ln/functional_tests.rs
Comment (jkczyz) [inline lightning/src/ln/functional_tests.rs:8699]: Is there a need to use a `Vec`? Would storing the last block header and height suffice?

**Review by jkczyz — COMMENTED** [review]
Need to take a closer look at the tests in the later commits.

File: lightning/src/ln/functional_tests.rs
Comment (TheBlueMatt) [reply lightning/src/ln/functional_tests.rs:1502]: Entirely, the `800_000` wasn't even right anymore - now that we subtract fees no output was ever exactly the 800k msat

File: lightning/src/util/test_utils.rs
Comment (TheBlueMatt) [reply lightning/src/util/test_utils.rs:205]: Hmm, not sure what the "Likewise" here is, but `TestBroadcaster` gets passed into various structs with a reference to the same blocks as the `Node` object holds. It failed to compile without being thread-safe (ie arc+mutex), but I didn't investigate further as it seemed sensible that it wouldn't work with an Rc.

File: lightning/src/chain/package.rs
Comment (TheBlueMatt) [reply lightning/src/chain/package.rs:290]: Actually, yea. I was avoiding doing it to keep the panic in debug, but I think you're right, if we ever *actually* get there, its probably indication that we're gonna generate an invalid spend transaction and/or burn funds to fee, which is not ok.

File: lightning/src/chain/onchaintx.rs
Comment (TheBlueMatt) [reply lightning/src/chain/onchaintx.rs]: Hmm, I'm not really sure this is at all more readable, but maybe I just fine long chains of iterators a bit harder to read. Thoughts?

```
				let locked_package_timelock = self.locktimed_packages.iter().map(|v| v.1.iter()).flatten()
					.find_map(|locked_package| if locked_package.outpoints() == req.outpoints() {
						Some(locked_package.package_timelock())
					} else { None });
				if let Some(timelock) = locked_package_timelock {
					log_trace!(logger, "Ignoring second claim for outpoint {}:{}, we already have one which we're waiting on a timelock at {} for.",
						req.outpoints()[0].txid, req.outpoints()[0].vout, timelock);
					continue;
				}
```

File: lightning/src/chain/onchaintx.rs
Comment (TheBlueMatt) [reply lightning/src/chain/onchaintx.rs:450]: No, I think you're very right - I was thking that because we don't have a `drain` we shouldn't as we'll end up copying a tone, but it doesn't actually matter because we can `drain` (or `append`-from) the inner `Vec`. It does make it much more readable.

File: lightning/src/ln/functional_test_utils.rs
Comment (TheBlueMatt) [reply lightning/src/ln/functional_test_utils.rs:215]: Ah, see https://github.com/rust-bitcoin/rust-lightning/pull/932#discussion_r641619700, but both the `Node` and the `TestBroadcaster` hold a reference to the same set of blocks.

File: lightning/src/chain/channelmonitor.rs
Comment (TheBlueMatt) [reply lightning/src/chain/channelmonitor.rs:1914]: Right, its just noting that we can always broadcast when the locktime is `X + 1`. I can drop if you think its obvious but Antoine suggested it.

* Fun fact of the day - Bitcoin Core had a bug for many years where it actually wouldn't accept-for-relay transactions until they were confirm-able in the *current* block, not X+1.

File: lightning/src/chain/onchaintx.rs
Comment (jkczyz) [reply lightning/src/chain/onchaintx.rs]: Given `find` returns an `Option`, you can use that instead of `find_map` with one line:

```rust
.find(|locked_package| locked_package.outpoints() == req.outpoints())
```

and keep the `log_trace` as it was.

Comment (TheBlueMatt) [reply lightning/src/chain/onchaintx.rs]: Oh! I misread the docs for `find`. Took your suggestion.

File: lightning/src/chain/package.rs
Comment (TheBlueMatt) [reply lightning/src/chain/package.rs]: I switched to an iterator as Jeff suggested below, so there is no variable name to be confused about :).

File: lightning/src/ln/functional_test_utils.rs
Comment (TheBlueMatt) [reply lightning/src/ln/functional_test_utils.rs:215]: Right, but the objects which own the broadcaster are thread-safe, so we have to be :/

File: lightning/src/ln/reorg_tests.rs
Comment (jkczyz) [inline lightning/src/ln/reorg_tests.rs]: Consider just cloning the entire `Vec` rather than holding a lock, given you're gonna clone some elements anyway.

File: lightning/src/ln/functional_tests.rs
Comment (jkczyz) [inline lightning/src/ln/functional_tests.rs:5845]: Could use `max` on an iterator here, too.

**Review by jkczyz — COMMENTED** [review]
Just a couple nits, but otherwise this is still outstanding: https://github.com/rust-bitcoin/rust-lightning/pull/932/commits/ff8ca028d538ec1e695aba814eb8be557a20544d#r641682315

File: lightning/src/chain/channelmonitor.rs
Comment (TheBlueMatt) [reply lightning/src/chain/channelmonitor.rs:1915]: Right, if the current height is X, then the next block which will be mined is X + 1. Thus, we can always broadcast something with a CLTV expiry of X + 1 as it can go into the next block (but could not appear in the current best block).

**Comment by TheBlueMatt** [comment]
Hmm, your link isn't working for me now, but I did try to respond at https://github.com/rust-bitcoin/rust-lightning/pull/932/commits/ff8ca028d538ec1e695aba814eb8be557a20544d#r641820784

File: lightning/src/chain/channelmonitor.rs
Comment (jkczyz) [reply lightning/src/chain/channelmonitor.rs:1915]: Ah, that makes sense. I think dropping "not the current height" would be beneficial as including it can lead to confusion.

**Review by jkczyz — APPROVED** [review]

**Review by ariard — COMMENTED** [review]
Code Review ACK a59b879

Changes since last time are addressing Jeff's comments

**Comment by TheBlueMatt** [comment]
Tweaked the comment as suggested at https://github.com/rust-bitcoin/rust-lightning/pull/932#discussion_r641829233, fixed one trivial error from new fixes, and squashed. Will merge after CI:

```
$ git diff-tree -U1 a59b8798 d51d606ba
diff --git a/lightning/src/chain/channelmonitor.rs b/lightning/src/chain/channelmonitor.rs
index 3df0111f5..26187ce34 100644
--- a/lightning/src/chain/channelmonitor.rs
+++ b/lightning/src/chain/channelmonitor.rs
@@ -1914,3 +1914,3 @@ impl<Signer: Sign> ChannelMonitorImpl<Signer> {
 					// Note that we add + 1 as transactions are broadcastable when they can be
-					// confirmed in the next block, not the current height.
+					// confirmed in the next block.
 					continue;
diff --git a/lightning/src/ln/functional_tests.rs b/lightning/src/ln/functional_tests.rs
index 30b9d0145..eb4f6054b 100644
--- a/lightning/src/ln/functional_tests.rs
+++ b/lightning/src/ln/functional_tests.rs
@@ -5843,3 +5843,3 @@ fn test_key_derivation_params() {
 	// Ensure all nodes are at the same height
-	let node_max_height = nodes.iter().map(|node| node.blocks.borrow().len()).max().unwrap() as u32;
+	let node_max_height = nodes.iter().map(|node| node.blocks.lock().unwrap().len()).max().unwrap() as u32;
 	connect_blocks(&nodes[0], node_max_height - nodes[0].best_block_info().1);
$
```

**Pull request merged**
